File: packages/nodejs/src/synthesized-function.ts
import { type z } from './flappy-type'
import { type SynthesizedFunctionDefinition } from './flappy-agent.interface'
import { type FlappyAgent } from './flappy-agent'
import { type ChatMLResponse, type ChatMLMessage } from './llm/interface'
import { FlappyFunctionBase, type FlappyFunctionOptions } from './flappy-function-base'
import { omit } from 'radash'

const extractSchema = (schema: any, prop: string): string =>
  JSON.stringify(omit(schema.parameters.properties[prop], ['description']))

export class SynthesizedFunction<
  TName extends string = string,
  TArgs extends z.ZodType = z.ZodType,
  TReturn extends z.ZodType = z.ZodType
> extends FlappyFunctionBase<TName, TArgs, TReturn> {
  declare define: SynthesizedFunctionDefinition<TName, TArgs, TReturn>

  public async call(
    agent: FlappyAgent,
    args: z.infer<TArgs>,
    options?: FlappyFunctionOptions
  ): Promise<z.infer<TReturn>> {
    const describe = this.define.description
    const returnTypeSchema = extractSchema(this.callingSchema, 'returnType')
    const argsSchema = extractSchema(this.callingSchema, 'args')
    const prompt = (args as any) instanceof Object ? JSON.stringify(args) : args
    const originalRequestMessage: ChatMLMessage[] = [
      {
        role: 'system',
        content: `${describe}
        User request according to the following JSON Schema:
        ${argsSchema}

        Translate it into JSON objects according to the following JSON Schema:
        ${returnTypeSchema}`
      },
      {
        role: 'user',
        content: `user request:${prompt}\n\njson object:`
      }
    ]
    let requestMessage = originalRequestMessage
    let retry = options?.retry ?? agent.retry
    let result: ChatMLResponse | undefined

    while (true) {
      try {
        if (retry !== agent.retry) console.debug('Attempt retry: ', agent.retry - retry)
        console.dir(requestMessage, { depth: null })
        result = await agent.llm.chatComplete(requestMessage)
        const data = this.parseComplete(result)
        return data
      } catch (err) {
        console.error(err)
        if (retry <= 0) throw new Error('Interrupted, function call failed. Please refer to the error message above.')

        retry -= 1
        // if the response came from chatComplete is failed, retry it directly.
        // Otherwise, update message for repairing
        if (result?.success && result.data) {
          requestMessage = [
            ...originalRequestMessage,
            {
              role: 'assistant',
              content: result?.data ?? ''
            },
            {
              role: 'user',
              content: `You response is invalid for the following reason:
          ${(err as Error).message}

          Please try again.`
            }
          ]
        }
      }
    }
  }

  protected parseComplete(resp: ChatMLResponse): string {
    const startIdx = resp.data!.indexOf('{')
    const endIdx = resp.data!.lastIndexOf('}')
    if (!(startIdx >= 0 && endIdx > startIdx)) throw new Error('Invalid JSON response')
    const json = JSON.parse(resp.data!.slice(startIdx, endIdx + 1))
    this.define.returnType.parse(json)
    console.debug(json)
    return json
  }
}

/**
 * Create an synthesized function.
 * @param define
 * @returns
 */
export const createSynthesizedFunction = <
  const TName extends string,
  const TArgs extends z.ZodType,
  const TReturn extends z.ZodType
>(
  define: SynthesizedFunctionDefinition<TName, TArgs, TReturn>
): SynthesizedFunction<TName, TArgs, TReturn> => new SynthesizedFunction(define)

File: packages/csharp/Pleisto.Flappy/SynthesizedFunction.cs
using Newtonsoft.Json.Linq;
using Pleisto.Flappy.Interfaces;
using Pleisto.Flappy.LLM.Interfaces;

namespace Pleisto.Flappy
{
  /// <summary>
  /// SynthesizedFunction
  /// </summary>
  /// <typeparam name="TArgs">Argument of functions</typeparam>
  /// <typeparam name="TReturn">Return of functions</typeparam>
  public class SynthesizedFunction<TArgs, TReturn> : FlappyFunctionBase<TArgs, TReturn>, IFlappyFunction
    where TArgs : class
    where TReturn : class
  {
    /// <summary>
    /// Function Call
    /// </summary>
    /// <param name="agent"></param>
    /// <param name="args"></param>
    /// <returns></returns>
    public override async Task<TReturn> Call(FlappyAgent agent, TArgs args)
    {
      var describe = Define.Description;
      var schema = FlappyAgent.GetSchemaGenerator();
      var returnTypeSchema = schema.Generate(typeof(TReturn)); //extractSchema(callingSchema, "returnType");
      var argsSchema = schema.Generate(typeof(TArgs)); ;
      var prompt = JObject.FromObject(args); //(args as object) is string ? args as string : JObject.FromObject(args).ToString();
      var requestMessage = new ChatMLMessage[]{
                  new ChatMLMessage
                  {
                    Role = ChatMLMessageRole.System,
                    Content = $@"{describe}
User reqeust according to the following JSON Schema:
${argsSchema}

Translate it into JSON objecvts according to the following JSON Schema:
{returnTypeSchema}"
                  },
                  new ChatMLMessage
                  {
                    Role = ChatMLMessageRole.User,
                    Content = $"user request:{prompt}\n\njson object:"
                  }
              };
      var result = await agent.llm.ChatComplete(requestMessage);
      try
      {
        var data = ParseComplete(result);

        return data;
      }
      catch (Exception ex)
      {
        var repaired = await agent.llm.ChatComplete(requestMessage.Union(new ChatMLMessage[]
        {
                    new ChatMLMessage{
                        Role = ChatMLMessageRole.Assistant,
                        Content= result.Data ?? ""
                    },
                    new ChatMLMessage{
                        Role = ChatMLMessageRole.User,Content=@$"Your response is invalid for the following reason:
${ex.Message}

Please try again."}
        }).ToArray(), null);
        return ParseComplete(repaired);
      }
    }

    private TReturn ParseComplete(ChatMLResponse msg)
    {
      var startIdx = msg.Data?.IndexOf('{') ?? -1;
      var endIdx = msg.Data?.LastIndexOf('}') ?? -1;
      if (startIdx == -1 || endIdx == -1 || endIdx < startIdx)
        throw new InvalidDataException($"Invalid JSON response startIndex={startIdx} endIdx={endIdx}");

      var content = msg.Data.Substring(startIdx, endIdx - startIdx + 1).Trim();
      try
      {
        var json = JObject.Parse(content);
        return json.ToObject<TReturn>();
      }
      catch (Exception ex)
      {
        throw new InvalidDataException($"exception on {nameof(ParseComplete)} {Environment.NewLine}{content}", ex);
      }
    }

    /// <summary>
    /// Create Synthesized Function
    /// </summary>
    /// <param name="define">Synthesized Definition</param>
    public SynthesizedFunction(SynthesizedFunctionDefinition<TArgs, TReturn> define) : base(define)
    {
    }
  }
}
